File: Advanced/PageAndRDL/SvgImage/VB.NET/SvgDesigner/SvgDesigner.vb
Imports System.Xml
Imports System.Runtime.InteropServices
Imports GrapeCity.ActiveReports.PageReportModel
Imports GrapeCity.ActiveReports.Design.DdrDesigner.Designers
Imports GrapeCity.ActiveReports.Design.DdrDesigner.Behavior
Imports System.Windows.Forms
Imports System.Drawing.Design
Imports System.Drawing
Imports System.ComponentModel.Design
Imports System.ComponentModel
Imports System.Drawing.Imaging
Imports GrapeCity.ActiveReports.Drawing.Gdi
Imports Svg


<Guid("ED362C95-D11C-4D98-ACF2-060F9A7FE68C")>
<ToolboxBitmap(GetType(SvgDesigner), "SvgIcon.png")>
<DisplayName("SvgImage")>
Public NotInheritable Class SvgDesigner
	Inherits CustomReportItemDesigner
	Implements IValidateable

	Private _designerVerbCollection As DesignerVerbCollection
	Public Sub New()
		AddProperty(Me, Function(x) x.Svg, CategoryAttribute.Data, New DescriptionAttribute(My.Resources.SvgDescription), New DisplayNameAttribute(My.Resources.SvgDisplayName), New EditorAttribute(GetType(MultilineStringEditor), GetType(UITypeEditor)))
	End Sub

	Protected Overrides Sub Dispose(disposing As Boolean)
		If disposing Then
			_graphics.Dispose()
			If _metafile IsNot Nothing Then _metafile.Dispose()
			_metafile = Nothing
		End If

		MyBase.Dispose(disposing)
	End Sub

	Function Validate(context As ValidationContext) As ValidationEntry() Implements IValidateable.Validate
		Return New ValidationEntry(0) {}
	End Function

#Region "Public properties"

	Public Overrides ReadOnly Property Verbs() As DesignerVerbCollection
		Get
			If IsNothing(_designerVerbCollection) Then
				_designerVerbCollection = New DesignerVerbCollection()
			End If
			Return _designerVerbCollection
		End Get
	End Property

	Public Property Svg() As String
		Get
			Dim customProperty = ReportItem.CustomProperties("Svg")
			If customProperty IsNot Nothing Then
				Dim expValue = customProperty.Value
				If expValue.IsConstant Then
					Return expValue.ToString()
				End If
			End If
			Return String.Empty
		End Get
		Set
			_metafile = Nothing
			Dim customProperty = ReportItem.CustomProperties("Svg")
			If customProperty IsNot Nothing Then
				ReportItem.CustomProperties.Remove(customProperty)
			End If
			customProperty = New CustomPropertyDefinition("Svg", Value)
			ReportItem.CustomProperties.Add(customProperty)
		End Set
	End Property

#End Region

#Region "Design-time rendering"

	Private ReadOnly _graphics As Graphics = SafeGraphics.CreateReferenceGraphics()
	Private _metafile As Metafile

	Friend ReadOnly Property RenderedSvg As Metafile
		Get
			If String.IsNullOrWhiteSpace(Svg) Then Return Nothing
			If _metafile IsNot Nothing Then Return _metafile
			Dim doc = New XmlDocument()
			doc.LoadXml(Svg)
			Dim document = SvgDocument.Open(doc)
			_metafile = New Metafile(_graphics.GetHdc(), (CType(document, ISvgBoundable)).Bounds, MetafileFrameUnit.Pixel, EmfType.EmfPlusDual)

			Try
				Using gfx = Graphics.FromImage(_metafile)
					document.Draw(gfx)
				End Using
			Finally
				_graphics.ReleaseHdc()
			End Try

			Return _metafile
		End Get
	End Property

	Private _controlGlyph As SvgControlGlyph

	Public Overrides ReadOnly Property ControlGlyph() As Glyph
		Get
			If (Not IsNothing(_controlGlyph)) Then
				Return _controlGlyph
			Else
				_controlGlyph = New SvgControlGlyph(ReportItem, Me)
				Return _controlGlyph
			End If
		End Get
	End Property

	Private NotInheritable Class SvgControlGlyph
		Inherits ControlBodyGlyph
		Public Sub New(reportItem As ReportItem, designer As SvgDesigner)
			MyBase.New(designer.BehaviorService, designer)
			Behavior = New MovableBehavior(reportItem, Me)
		End Sub

		Public Overrides Sub Paint(pe As PaintEventArgs)
			If PaintSelectionOnly(pe) Then
				MyBase.Paint(pe)
				Return
			End If

			pe.Graphics.FillRectangle(BackgroundBrush, Bounds)
			Try
				Dim svg = (CType(ComponentDesigner, SvgDesigner)).RenderedSvg
				If svg IsNot Nothing Then pe.Graphics.DrawImage(svg, Bounds)
			Catch ex As Exception
				pe.Graphics.DrawString(ex.Message, SystemFonts.DefaultFont, SystemBrushes.ControlText, Bounds, StringFormat.GenericTypographic)
			End Try

			MyBase.Paint(pe)
		End Sub

		Private NotInheritable Class MovableBehavior
			Inherits DefaultControlBehavior
			Private ReadOnly _reportItem As ReportItem
			Private ReadOnly _glyph As Glyph

			Public Sub New(reportItem As ReportItem, glyph As Glyph)
				_reportItem = reportItem
				_glyph = glyph
			End Sub

			Public Overrides ReadOnly Property Cursor() As Cursor
				Get
					Return If(CanMoveGlyph(_glyph) AndAlso IsActiveLayerItem(_reportItem), Cursors.SizeAll, MyBase.Cursor)
				End Get
			End Property

			Private Shared Function IsActiveLayerItem(item As ReportItem) As Boolean
				If item Is Nothing OrElse item.Site Is Nothing Then
					Return True
				End If
				Dim host = TryCast(item.Site.GetService(GetType(IDesignerHost)), IDesignerHost)
				If host Is Nothing Then
					Return True
				End If
				Dim reportDef = TryCast(host.RootComponent, PageReport)
				If reportDef Is Nothing Then
					Return True
				End If
				Dim reportDesigner = TryCast(host.GetDesigner(reportDef.Report), ReportDesigner)
				If reportDesigner Is Nothing Then
					Return True
				End If
				Dim activeLayer = reportDesigner.ActiveLayer
				Return String.Equals(activeLayer, item.LayerName, StringComparison.InvariantCultureIgnoreCase)
			End Function
		End Class
	End Class

#End Region
End Class

File: Advanced/PageAndRDL/SvgImage/C#/SvgDesigner/SvgDesigner.cs
using System;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Drawing;
using System.Drawing.Design;
using System.Drawing.Imaging;
using System.IO;
using System.Windows.Forms;
using GrapeCity.ActiveReports.Design.DdrDesigner.Behavior;
using GrapeCity.ActiveReports.Design.DdrDesigner.Designers;
using GrapeCity.ActiveReports.PageReportModel;
using System.Runtime.InteropServices;
using System.Xml;
using GrapeCity.ActiveReports.Drawing.Gdi;
using Svg;

namespace GrapeCity.ActiveReports.Samples.Svg
{
	[Guid("ED362C95-D11C-4D98-ACF2-060F9A7FE68C")]
	[ToolboxBitmap(typeof(SvgDesigner), "SvgIcon.png")]
	[DisplayName("SvgImage")]
	public sealed class SvgDesigner : CustomReportItemDesigner, IValidateable
	{
		private DesignerVerbCollection _designerVerbCollection;

		public SvgDesigner()
		{
			AddProperty(this, x => x.Svg, CategoryAttribute.Data,
				new DescriptionAttribute(Properties.Resources.SvgDescription), new DisplayNameAttribute(Properties.Resources.SvgDisplayName),
				new EditorAttribute(typeof(MultilineStringEditor), typeof(UITypeEditor))
			);
		}

		protected override void Dispose(bool disposing)
		{
			if (disposing)
			{
				_graphics.Dispose();
				if (_metafile != null)
					_metafile.Dispose();
				_metafile = null;
			}
			base.Dispose(disposing);
		}

		ValidationEntry[] IValidateable.Validate(ValidationContext context)
		{
			return new ValidationEntry[0];
		}

		public override DesignerVerbCollection Verbs
		{
			get { return _designerVerbCollection ?? (_designerVerbCollection = new DesignerVerbCollection()); }
		}

		#region Public properties

		public string Svg
		{
			get
			{
				var customProperty = ReportItem.CustomProperties["Svg"];
				if (customProperty != null)
				{
					var expValue = customProperty.Value;
					if (expValue.IsConstant)
						return expValue.ToString();
				}
				return string.Empty;
			}
			set
			{
				_metafile = null;
				var customProperty = ReportItem.CustomProperties["Svg"];
				if (customProperty != null)
					ReportItem.CustomProperties.Remove(customProperty);
				customProperty = new CustomPropertyDefinition("Svg", value);
				ReportItem.CustomProperties.Add(customProperty);
			}
		}

		#endregion

		#region Design-time rendering

		private readonly Graphics _graphics = SafeGraphics.CreateReferenceGraphics();
		private Metafile _metafile;
		
		internal Metafile RenderedSvg
		{
			get
			{
				if (string.IsNullOrWhiteSpace(Svg))
					return null;

				if (_metafile != null)
					return _metafile;
				
				var doc = new XmlDocument();
				doc.LoadXml(Svg);
				var svgDocument = SvgDocument.Open(doc);
					
				_metafile = new Metafile(_graphics.GetHdc(), ((ISvgBoundable)svgDocument).Bounds, MetafileFrameUnit.Pixel, EmfType.EmfPlusDual);
				try
				{
					using (var gfx = Graphics.FromImage(_metafile))
						svgDocument.Draw(gfx);
				}
				finally
				{
					_graphics.ReleaseHdc();
				}
				return _metafile;
			}
		}

		private SvgControlGlyph _controlGlyph;

		public override Glyph ControlGlyph
		{
			get { return _controlGlyph ?? (_controlGlyph = new SvgControlGlyph(ReportItem, this)); }
		}

		private sealed class SvgControlGlyph : ControlBodyGlyph
		{
			public SvgControlGlyph(ReportItem reportItem, SvgDesigner designer)
				: base(designer.BehaviorService, designer)
			{
				Behavior = new MovableBehavior(reportItem, this);
			}

			public override void Paint(PaintEventArgs pe)
			{
				if (PaintSelectionOnly(pe))
				{
					base.Paint(pe);
					return;
				}

				pe.Graphics.FillRectangle(BackgroundBrush, Bounds);
				try
				{
					var svg = ((SvgDesigner) ComponentDesigner).RenderedSvg;
					if (svg != null)
						pe.Graphics.DrawImage(svg, Bounds);
				}
				catch (Exception ex)
				{
					pe.Graphics.DrawString(ex.Message, SystemFonts.DefaultFont, SystemBrushes.ControlText, Bounds, StringFormat.GenericTypographic);
				}

				base.Paint(pe);
			}

			private sealed class MovableBehavior : DefaultControlBehavior
			{
				private readonly ReportItem _reportItem;
				private readonly Glyph _glyph;

				public MovableBehavior(ReportItem reportItem, Glyph glyph)
				{
					_reportItem = reportItem;
					_glyph = glyph;
				}

				public override Cursor Cursor
				{
					get
					{
						return CanMoveGlyph(_glyph) && IsActiveLayerItem(_reportItem) ? Cursors.SizeAll : base.Cursor;
					}
				}

				private static bool IsActiveLayerItem(ReportItem item)
				{
					if (item == null || item.Site == null)
						return true;
					var host = item.Site.GetService(typeof(IDesignerHost)) as IDesignerHost;
					if (host == null)
						return true;
					var reportDef = host.RootComponent as PageReport;
					if (reportDef == null)
						return true;
					var reportDesigner = host.GetDesigner(reportDef.Report) as ReportDesigner;
					if (reportDesigner == null)
						return true;
					var activeLayer = reportDesigner.ActiveLayer;
					return string.Equals(activeLayer, item.LayerName, StringComparison.InvariantCultureIgnoreCase);
				}
			}
		}

		#endregion
	}

}
